Validate hostname starting from 1.24 API.

vdemeester opened the pull request:
In order to keep a little bit of "sanity" on the API side, validate hostname only starting from v1.24 API version 🐯.

Fixes #23729.

/cc @thaJeztah @tiborvass @icecrime 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

thaJeztah commented:
:heart: thank you!

/cc @ibuildthecloud

icecrime commented:
Cc @alena1108: works for you?

alena1108 commented:
@icecrime yup, thank you!

thaJeztah commented:
LGTM

tiborvass commented:
I'm not fond of passing version around in the API :S not sure what the alternative is...

runcom commented:
can't we have a bool argument instead?
```
version := httputils.VersionFromContext(ctx)
validateHostname := versions.GreaterThanOrEqualTo(version, "1.24")
```
and then pass around `validateHostname`

vdemeester commented:
That's a good idea :wink:

vdemeester commented:
Rebased and update :angel: 
@runcom @tiborvass wdyt ?

runcom commented:
:+1: LGTM

tiborvass commented:
@vdemeester TestParseWords is failing

vdemeester commented:
@tiborvass hum… I'll look into it :angel:

vdemeester commented:
@tiborvass rebase because I don't reproduce the failure of this test :sweat: (which is worrying me a bit :sob: ).

tiborvass commented:
Much better! LGTM

thaJeztah commented:
relates to https://github.com/docker/docker/pull/21641, https://github.com/docker/docker/issues/21595

thaJeztah commented:
@vdemeester can you add a mention of this change to https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v124-api-changes ?

thaJeztah commented:
Perhaps we should mention it in the "HostName" JSON option in https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.24.md#create-a-container as well (that it must be a valid  RFC 1123 hostname, per https://github.com/docker/docker/pull/20566)

vdemeester committed:
Validate hostname starting from 1.24 API.

In order to keep a little bit of "sanity" on the API side, validate
hostname only starting from v1.24 API version.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
@thaJeztah updated 👼

thaJeztah commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -133,9 +133,10 @@ This section lists each version from latest to oldest.  Each listing includes a
 * API errors are now returned as JSON instead of plain text.
 * `POST /containers/create` and `POST /containers/(id)/start` allow you to configure kernel parameters (sysctls) for use in the container.
-* `POST /v1.23/containers/<container ID>/exec` and `POST /v1.23/exec/<exec ID>/start`
+* `POST /containers/<container ID>/exec` and `POST /exec/<exec ID>/start`
```
Thx :blush:

vdemeester replied:
:stuck_out_tongue_closed_eyes:

thaJeztah commented:
LGTM, thanks!
